x-pack/filebeat/input/netflow/input.go

```diff
@@ -34,33 +35,176 @@ var (
+		f, err := LoadFieldDefinitionsFromFile(yamlPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed parsing custom field definitions from file '%s': %w", yamlPath, err)
+		}
+		customFields[idx] = f
+	}
+
+	dec, err := decoder.NewDecoder(decoder.NewConfig().
+		WithProtocols(inputCfg.Protocols...).
+		WithExpiration(inputCfg.ExpirationTimeout).
+		WithLogOutput(&logDebugWrapper{Logger: im.log}).
+		WithCustomFields(customFields...).
+		WithSequenceResetEnabled(inputCfg.DetectSequenceReset).
+		WithSharedTemplates(inputCfg.ShareTemplates))
+	if err != nil {
+		return nil, fmt.Errorf("error initializing netflow decoder: %w", err)
+	}
+
+	input := &netflowInput{
+		decoder:          dec,
+		internalNetworks: inputCfg.InternalNetworks,
+		logger:           im.log,
+		queueSize:        inputCfg.PacketQueueSize,
+	}
+
+	input.udp = udp.New(&inputCfg.Config, input.packetDispatch)
+
+	return input, nil
+}
+
 type packet struct {
 	data   []byte
 	source net.Addr
 }
 
 type netflowInput struct {
-	mutex            sync.Mutex
+	mtx              sync.Mutex
 	udp              *udp.Server
 	decoder          *decoder.Decoder
-	outlet           channel.Outleter
-	forwarder        *harvester.Forwarder
+	client           beat.Client
 	internalNetworks []string
 	logger           *logp.Logger
 	queueC           chan packet
 	queueSize        int
 	started          bool
 }
 
-func init() {
-	err := input.Register(inputName, NewInput)
+func (n *netflowInput) Name() string {
+	return inputName
+}
+
+func (n *netflowInput) Test(_ v2.TestContext) error {
+	return nil
+}
+
+func (n *netflowInput) packetDispatch(data []byte, metadata inputsource.NetworkMetadata) {
+	select {
+	case n.queueC <- packet{data, metadata.RemoteAddr}:
+		numPackets.Inc()
+	default:
+		numDropped.Inc()
+	}
+}
+
+func (n *netflowInput) Run(context v2.Context, connector beat.PipelineConnector) error {
+	n.mtx.Lock()
+	if n.started {
+		n.mtx.Unlock()
+		return nil
+	}
+
+	n.started = true
+	n.mtx.Unlock()
+
+	n.logger.Info("Starting netflow input")
+
+	n.logger.Info("Connecting to beat event publishing")
+	client, err := connector.ConnectWith(beat.ClientConfig{
+		PublishMode: beat.DefaultGuarantees,
+		Processing: beat.ProcessingConfig{
+			// This input only produces events with basic types so normalization
+			// is not required.
+			EventNormalization: boolPtr(false),
+		},
+		CloseRef:      context.Cancelation,
+		EventListener: nil,
+	})
+	if err != nil {
+		n.logger.Errorw("Failed connecting to beat event publishing", "error", err)
+		n.stop()
+		return err
+	}
+
+	n.logger.Info("Starting netflow decoder")
+	if err := n.decoder.Start(); err != nil {
+		n.logger.Errorw("Failed to start netflow decoder", "error", err)
+		n.stop()
+		return err
+	}
+
+	n.queueC = make(chan packet, n.queueSize)
+
+	n.logger.Info("Starting udp")
+	err = n.udp.Start()
 	if err != nil {
-		panic(err)
+		n.logger.Errorf("Failed to start udp: %v", err)
+		n.stop()
+		return err
+	}
```

Comment:
It's a small detail, however I believe it would make the log/error message more clear if we explicitly mention it's a UDP server.
```suggestion
	n.logger.Info("Starting udp server")
	err = n.udp.Start()
	if err != nil {
		n.logger.Errorf("Failed to start udp server: %v", err)
		n.stop()
		return err
	}
```